File: binding-python/runtime/src/main/python/etch/binding/transport/PlainMailboxManager.py
"""
#  Licensed to the Apache Software Foundation (ASF) under one   *
#  or more contributor license agreements.  See the NOTICE file *
#  distributed with this work for additional information        *
#  regarding copyright ownership.  The ASF licenses this file   *
#  to you under the Apache License, Version 2.0 (the            *
#  "License"); you may not use this file except in compliance   *
#  with the License.  You may obtain a copy of the License at   *
#                                                               *
#    http://www.apache.org/licenses/LICENSE-2.0                 *
#                                                               *
#  Unless required by applicable law or agreed to in writing,   *
#  software distributed under the License is distributed on an  *
#  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
#  KIND, either express or implied.  See the License for the    *
#  specific language governing permissions and limitations      *
#  under the License. 
"""
from __future__ import absolute_import
from __future__ import with_statement
from ...python.SynchronizedOn import *
from ...python.SynchronizedMap import *
from ...python.Exceptions import *
from .MailboxManager import *
from ..msg.Message import *
from ..support.Mailbox import *
from ...util.IdGenerator import *
from ...util.Resources import *
from ...util.URL import *
from ...util.core.Who import *
from ...util.core.io.Session import *

class PlainMailboxManager(MailboxManager):
    """
    MailboxManager is a MessageHandler which accepts packets for possible delivery to a mailbox, or
    to another message handler if an appropriate mailbox cannot be found. MailboxManager is also a
    MessageSource, tagging messages with msgIds and then forwarding them to another MessageSource.
    If requested, a mailbox is created with a message's msgid and added to the set of mailboxes waiting
    for messages.
    """
    
    def __init__(self, transport, uri, resources):
        """
        @param transport used to deliver messages to our peer
        @param uri
        @param resources
        """
        self.__transport = transport
        self.__session   = None
        self.__up        = False
        self.__mailboxes = SynchronizedMap()
        
        # TODO - change 'seed' 
        self.__idGen     = IdGenerator('seed', 37)
        self.__transport.setSession(self)
    
    def getTransport(self):
        """
        @return the transport
        """
        return self.__transport
    
    def __repr__(self):
        return "PlainMailboxManager/%s" % repr(self.__transport)
    
    ####
    # MessageHandler Methods
    ####
    
    def sessionMessage(self, sender, msg):
        msgid = msg.getInReplyTo()
        if msgid != None:
            mb = self.getMailbox(msgid)
            if mb != None:
                try:
                    return mb.message(sender, msg)
                except InterruptedException:
                    # timeout or mailbox closed .. fall through
                    pass
            # no such mailbox .. reject
            return False
        # no msgid .. pass of to session
        return self.__session.sessionMessage(sender, msg)
    
    ####
    # MessageSource Methods
    ####
    
    def transportMessage(self, recipient, msg):
        if msg.getMessageId() != None:
            raise IllegalStateException, "message has already been sent"
        
        msg.setMessageId(self.__idGen.next())
        
        self.__transport.transportMessage(recipient, msg)
    
    def transportCall(self, recipient, msg):
        if msg.getMessageId() != None:
            raise IllegalStateException, "message has already been sent"
        
        if msg.getInReplyTo() != None:
            raise IllegalStateException, "message is marked as a reply"
        
        msgid = self.__idGen.next()
        msg.setMessageId(msgid)
        
        mb = PlainMailbox(self, msgid)
        self.register(mb)
        
        try:
            self.__transport.transportMessage(recipient, msg)
        except:
            self.unregister(mb)
            raise Exception
        
        return mb
    
    ####
    # Mailbox methods
    ####
    
    def register(self, mb):
        """
        Adds a mailbox to the set of mailbox receiving responses to messages.
        @param mb
        """
        msgid = mb.getMessageId()
        with SynchronizedOn(self.__mailboxes):
            if not self.__up:
                raise IllegalStateException, "connection down"
            
            if msgid in self.__mailboxes.keys():
                raise IllegalArgumentException, "duplicate msgid in mailboxes"
            
            self.__mailboxes[msgid] = mb
        
        
    def unregister(self, mb):
        del self.__mailboxes[mb.getMessageId()]
    
    def unregisterAll(self):
        for mb in self.__mailboxes.values():
            mb.closeDelivery()
    
    def getMailbox(self, msgid):
        """
        Returns the mailbox for the specified msgid. *This is a testing api*
        @param msgid
        @return the mailbox for the specified msgid
        """
        return self.__mailboxes[msgid]
    
    def redeliver(self, sender, msg):
        self.__session.sessionMessage(sender, msg)
    
    def size(self):
        """
        @return the number of mailboxes
        """
        return len(self.__mailboxes)
    
    ####
    # SourceHandler methods
    ####
    
    def sessionQuery(self, query):
        return self.__session.sessionQuery(query)
    
    def sessionControl(self, control, value):
        self.__session.sessionControl(control, value)
    
    def sessionNotify(self, event):
        if event == Session.UP:
            self.__up = True
        elif event == Session.DOWN:
            self.__up = False
            self.unregisterAll()
        self.__session.sessionNotify(event)
    
    ####
    # Source methods
    ####
    
    def transportQuery(self, query):
        return self.__transport.transportQuery(query)
    
    def transportControl(self, control, value):
        self.__transport.transportControl(control, value)
    
    def transportNotify(self, event):
        self.__transport.transportNotify(event)
    
    def getSession(self):
        return self.__session
    
    def setSession(self, session):
        self.__session = session


        
    
    
    
    

File: binding-java/runtime/src/main/java/org/apache/etch/bindings/java/transport/PlainMailboxManager.java
/* $Id$
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements. See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership. The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License. You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package org.apache.etch.bindings.java.transport;

import java.util.Collections;
import java.util.HashMap;
import java.util.Map;

import org.apache.etch.bindings.java.msg.Message;
import org.apache.etch.bindings.java.support.Mailbox;
import org.apache.etch.util.IdGenerator;
import org.apache.etch.util.Resources;
import org.apache.etch.util.URL;
import org.apache.etch.util.core.Who;
import org.apache.etch.util.core.io.Session;


/**
 * MailboxManager is a MessageHandler which accepts packets for
 * possible delivery to a mailbox, or to another message handler
 * if an appropriate mailbox cannot be found. MailboxManager is
 * also a MessageSource, tagging messages with msgids and then
 * forwarding them to another MessageSource. If requested, a
 * mailbox is created with a message's msgid and added to the
 * set of mailboxes waiting for messages.
 */
public class PlainMailboxManager implements MailboxManager
{
	/**
	 * @param transport used to deliver messages to our peer.
	 * @param uri
	 * @param resources
	 */
	public PlainMailboxManager( TransportMessage transport, String uri,
		Resources resources )
	{
		this( transport, new URL( uri ), resources );
	}

	/**
	 * @param transport used to deliver messages to our peer.
	 * @param uri
	 * @param resources
	 */
	public PlainMailboxManager( TransportMessage transport, URL uri,
		Resources resources )
	{
		this.transport = transport;
		transport.setSession( this );
	}
	
	private final TransportMessage transport;
	
	/**
	 * @return the transport.
	 */
	public TransportMessage getTransport()
	{
		return transport;
	}
	
	@Override
	public String toString()
	{
		return String.format( "PlainMailboxManager/%s", transport );
	}

	////////////////////////////
	// MessageHandler methods //
	////////////////////////////
	
	public boolean sessionMessage( Who sender, Message msg ) throws Exception
	{
		//Log.report( "MailboxManager.recv", "msg", msg );
		Long msgid = msg.getInReplyTo();
		if (msgid != null)
		{
			Mailbox mb = getMailbox( msgid );
			//Log.report( "MailboxManager.recv", "msg", msg, "mb", mb );
			if (mb != null)
			{
				try
				{
					return mb.message( sender, msg );
				}
				catch ( InterruptedException e )
				{
					// timeout or mailbox closed - fall through
				}
			}
			// no such mailbox - reject
			return false;
		}
		// no msgid - pass off to session
		return session.sessionMessage( sender, msg );
	}

	///////////////////////////
	// MessageSource methods //
	///////////////////////////

	public void transportMessage( Who recipient, Message msg ) throws Exception
	{
		if (msg.getMessageId() != null)
			throw new IllegalStateException( "message has already been sent" );
		
		msg.setMessageId( idGen.next() );
		
		//Log.report( "MailboxManager.send", "msg", msg );
		
		transport.transportMessage( recipient, msg );
	}

	public Mailbox transportCall( Who recipient, Message msg ) throws Exception
	{
		if (msg.getMessageId() != null)
			throw new IllegalStateException( "message has already been sent" );
		
		if (msg.getInReplyTo() != null)
			throw new IllegalStateException( "message is marked as a reply" );
		
		Long msgid = idGen.next();
		msg.setMessageId( msgid );
		
		Mailbox mb = new PlainMailbox( this, msgid );
		register( mb );
		
		//Log.report( "MailboxManager.send", "msg", msg );

		try
		{
			transport.transportMessage( recipient, msg );
		}
		catch ( Exception e )
		{
			unregister( mb );
			throw e;
		}
		
		return mb;
	}
	
	private final IdGenerator idGen = new IdGenerator( System.currentTimeMillis(), 37 );

	/////////////////////
	// Mailbox methods //
	/////////////////////
	
	/**
	 * Adds a mailbox to the set of mailbox receiving responses
	 * to messages.
	 * @param mb
	 */
	public void register( Mailbox mb )
	{
		Long msgid = mb.getMessageId();
		synchronized (mailboxes)
		{
			if (!up)
				throw new IllegalStateException( "connection down" );
			
			if (mailboxes.containsKey( msgid ))
				throw new IllegalArgumentException( "dup msgid in mailboxes" );
			
			mailboxes.put( msgid, mb );
		}
	}

	public void unregister( Mailbox mb )
	{
		mailboxes.remove( mb.getMessageId() );
	}
	
	private void unregisterAll()
	{
		for (Mailbox mb: mailboxes.values().toArray( new Mailbox[0] ))
			mb.closeDelivery();
	}
	
	/**
	 * Returns the mailbox for the specified msgid. This is a testing api.
	 * @param msgid
	 * @return the mailbox for the specified msgid.
	 */
	public Mailbox getMailbox( Long msgid )
	{
		return mailboxes.get( msgid );
	}
	
	private Map<Long, Mailbox> mailboxes = Collections.synchronizedMap(
		new HashMap<Long, Mailbox>() );

	public void redeliver( Who sender, Message msg ) throws Exception
	{
		session.sessionMessage( sender, msg );
	}

	/**
	 * @return the number of mailboxes
	 */
	public int size()
	{
		return mailboxes.size();
	}
	
	///////////////////////////
	// SourceHandler methods //
	///////////////////////////

	public Object sessionQuery( Object query ) throws Exception
	{
		return session.sessionQuery( query );
	}

	public void sessionControl( Object control, Object value ) throws Exception
	{
		session.sessionControl( control, value );
	}

	public void sessionNotify( Object event ) throws Exception
	{
		if (event == Session.UP)
		{
			up = true;
		}
		else if (event == Session.DOWN)
		{
			up = false;
			unregisterAll();
		}
		
		session.sessionNotify( event );
	}
	
	private boolean up;

	////////////////////
	// Source methods //
	////////////////////

	public Object transportQuery( Object query ) throws Exception
	{
		return transport.transportQuery( query );
	}

	public void transportControl( Object control, Object value ) throws Exception
	{
		transport.transportControl( control, value );
	}

	public void transportNotify( Object event ) throws Exception
	{
		transport.transportNotify( event );
	}

	public SessionMessage getSession()
	{
		return session;
	}
	
	public void setSession( SessionMessage session )
	{
		this.session = session;
	}
	
	private SessionMessage session;
}
